src/ripple/basics/CompressionAlgorithms.h

```diff
@@ -0,0 +1,272 @@
+
+    return originalSizeBytes + compressedSize;
+}
+
+/** Copy data from the input stream.
+ * @tparam InputStream ZeroCopyInputStream
+ * @param in Source input stream
+ * @param dst Destination buffer
+ * @param size Destination size
+ * @param usedSize Vector of consumed input chunk sizes
+ * @return true if the stream has enough data to copy
+ */
+template<typename InputStream>
+bool
+copyStream(InputStream& in, std::uint8_t* dst, int size, std::vector<int> &usedSize)
+{
+    const void * data = NULL;
+    int dataSize = 0;
+    int copied = 0;
+
+    while (copied != size && in.Next(&data, &dataSize))
+    {
+        auto sz = dataSize >= (size - copied) ? (size - copied) : dataSize;
+        std::memcpy(dst + copied, data, sz);
+        copied += sz;
+        usedSize.push_back(dataSize);
+    }
+
+    return copied == size;
+}
+
+/** Get the size of uncompressed data from the varint header.
+ * @tparam InputStream ZeroCopyInputStream
+ * @param in Source input stream
+ * @return Size of the uncompressed data
+ */
+template<typename InputStream>
+std::size_t
+getOriginalSize(InputStream& in)
+{
+    std::array<std::uint8_t, varint_traits<std::uint32_t>::max> vi;
+    std::size_t originalSize = 0;
+    const void * data = NULL;
+    int dataSize = 0;
+    std::vector<int> usedSize;
+
+    if (!in.Next(&data, &dataSize))
+        doThrow("lz4f decompress: invalid input size");
+
+    usedSize.push_back(dataSize);
+
+    const void *p = data;
+
+    if (dataSize < varint_traits<std::uint32_t>::max)
+    {
+        std::memcpy(vi.data(), data, dataSize);
+        if (!copyStream(in, vi.data() + dataSize,
+                        varint_traits<std::uint32_t>::max - dataSize, usedSize))
+            doThrow("lz4f decompress: header");
+        p = vi.data();
+    }
+
+    auto originalSizeBytes = read_varint(p, varint_traits<std::uint32_t>::max, originalSize);
+
+    if (originalSizeBytes == 0)
+        doThrow("lz4f decompress:: originalSizeBytes == 0");
+
+    // rewind the stream
+    for (auto it = usedSize.rbegin(); it != usedSize.rend(); ++it)
+        in.BackUp(*it);
+    in.Skip(originalSizeBytes);
+
+    return originalSize;
+}
+
+/** Read next available compressed chunk. If the buffer of cached compressed data is not empty then the chunk
+ * is copied into the buffer.
+ * @tparam InputStream ZeroCopyInputStream
+ * @param in Source input stream
+ * @param inSize Remaining size of the compressed data to read from the input stream.
+ * @param compressedChunk Pointer to the next chunk
+ * @param chunkSize Size of the next chunk
+ * @param buffer Cached compressed data
+ * @return true if the stream is not empty
+ */
+template<typename InputStream>
+bool
+nextChunk(InputStream &in, std::size_t inSize,
+          void const *& compressedChunk, int &chunkSize,
+          std::vector<std::uint8_t> &buffer)
+{
+    bool res = in.Next(&compressedChunk, &chunkSize);
+
+    if (!res && buffer.size() == 0)
```

Comment:
I think that `!buffer.empty()` reads better both here and on line 167.

Reply:
updated